Std/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,533 @@
+  pure <| (← importTree.getMatch ty config).reverse
+
+end DiscrTreeFinder
+
+namespace IncDiscrTreeFinder
+
+
+/--
+The maximum number of constants an individual task performed.
+
+The value below was picked because it roughly correponded to 50ms of work on the machine this was
+developed on.  Smaller numbers did not seem to improve performance when importing Std and larger
+numbers (<10k) seemed to degrade initialization performance.
+-/
+private def constantsPerTask : Nat := 6500
+
+private def addImport (name : Name) (constInfo : ConstantInfo) :
+    MetaM (Array (InitEntry (Name × DeclMod))) :=
+  forallTelescope constInfo.type fun _ type => do
+    let e ← InitEntry.fromExpr type (name, DeclMod.none)
+    let a := #[e]
+    let biff := e.key == DiscrTree.Key.const ``Iff 2
+    if biff then
+      let a := a.push (←e.mkSubEntry 0 (name, DeclMod.mp))
+      let a := a.push (←e.mkSubEntry 1 (name, DeclMod.mpr))
+      pure a
+    else
+      pure a
+
+/--
+Candidate finding function that uses strict discrimination tree for resolution.
+-/
+def mkImportFinder : IO CandidateFinder := do
+  let ref ← IO.mkRef none
+  pure fun ty => do
+    let importTree ← (←ref.get).getDM $ do
+      profileitM Exception  "librarySearch launch" (←getOptions) $
+        createImportedEnvironment (←getEnv) (constantsPerTask := constantsPerTask) addImport
+    let (imports, importTree) ← importTree.getMatch ty
+    ref.set importTree
+    pure imports
+
+end IncDiscrTreeFinder
+
+private unsafe def mkImportFinder : IO CandidateFinder := do
+  let path ← DiscrTreeFinder.cachePath
+  if ← path.pathExists then
+    let (imports, _) ← unpickle (DiscrTree (Name × DeclMod)) path
+    -- `DiscrTree.getMatch` returns results in batches, with more specific lemmas coming later.
+    -- Hence we reverse this list, so we try out more specific lemmas earlier.
+    pure <| DiscrTreeFinder.mkImportFinder {} imports
+  else do
+    IncDiscrTreeFinder.mkImportFinder
+
+
+/--
+Retrieve the current current of lemmas.
+-/
+initialize defaultCandidateFinder : IO.Ref CandidateFinder ← unsafe do
+  IO.mkRef (←mkImportFinder)
+
+/--
+Update the candidate finder used by library search.
+-/
+def setDefaultCandidateFinder (cf : CandidateFinder) : IO Unit :=
+  defaultCandidateFinder.set cf
+
+private def emoji (e:Except ε α) := if e.toBool then checkEmoji else crossEmoji
+
+/-- Create lemma from name and mod. -/
+def mkLibrarySearchLemma (lem : Name) (mod : DeclMod) : MetaM Expr := do
+  let lem ← mkConstWithFreshMVarLevels lem
+  match mod with
+  | .none => pure lem
+  | .mp => mapForallTelescope (fun e => mkAppM ``Iff.mp #[e]) lem
+  | .mpr => mapForallTelescope (fun e => mkAppM ``Iff.mpr #[e]) lem
+
+/--
+A library search candidate using symmetry includes the goal to solve, the metavar
+context for that goal, and the name and orientation of a rule to try using with goal.
+-/
+@[reducible]
+def Candidate :=  (MVarId × MetavarContext) × (Name × DeclMod)
```

Comment:
I'm confused why the `MVarId × MetavarContext` is here. These are always the same, as the second factor varies.

Reply:
Oh, I see, because of the way you're handling `symm` now, half the time it will be the `symm` of the goal.